### ISSUE TITLE
Questions on feature importance plot on a multi-class scenario

### ISSUE THREAD

**Issue by ivan-marroquin** [shap/shap#1029]
Hi all,

I have a trained classifier on multi-class using XGBClassifier from xgboost. 

I followed the code in: https://github.com/slundberg/shap/blob/master/notebooks/tree_explainer/XGBoost%20Multi-class%20Example.ipynb

to generate a feature importance plot (see attached zip file). My questions are the following:
1) I noticed that the horizontal bars in the feature importance plot are colored according to class labels. How do I interpret the portion assigned to a class in each of the horizontal bars

2) In the notebook mentioned above the validation set was used to compute shapley values. While in other notebooks, either the enitre data or training data is used. So, what is a better approach to compute the shapley values? Do I use the training set or is it better to use the validation set? In the attached figures, the ordering of features in different according to what data set was used.

Many thanks,

Ivan

[multi-class_feature_importance_plot.zip](https://github.com/slundberg/shap/files/4149862/multi-class_feature_importance_plot.zip)

**Comment by slundberg** [comment]
1. Each color of each bar is `np.abs(shap_values[j][:,i]).mean()` for feature i and output j/

2. It really depends on which dataset you want to use to evaluate your model. If I evaluate my model on a dataset where one feature is just set to its mean value for all samples then that feature will have little impact in that dataset, while for another dataset where that feature varies a lot it might matter more. If you see a difference between the global importance of a feature in the train vs. validation datasets it is either from random sampling differences between the datasets (likely if they are small datasets), or from a real distributional shift that is effecting your model.

**Comment by ivan-marroquin** [comment]
I think that I got - please confirm

1. In a multi-class classification, the shapley values are organized in a list of tensors (# classes x # samples x # features). The colors for each class in each bar denote the class interaction for a specific feature: sort of how many times that feature was used to pick the given class

2. The original train data was split into training set and validation set. The training set went undergo oversampling to correct for imbalanced class distribution. While the validation set was left as it is. The processing applied to the training set should explain the difference in the summary result plots. Therefore, the best option is to use the training set to determine feature importance

**Comment by slundberg** [comment]
1. Yes.
2. Yes, but both are fine. The validation will tell you how important each feature is relative to the real background rates (which is what you would expect in any future test examples). If you use the training ones, just remember that a 0 SHAP value means "uninformative" in the sense that it would not change anything to replace that value with what you expect based on some background prior distribution. You typically want that prior distribution to match the distribution of features in the real world, but if you have feature changes from the training distribution also means that some features might look more important in the validation set just because they all have "surprising values" and the SHAP values are hence not mean centered across the set.
